### DISCUSSION TITLE
How to add or remove one TreeViewItem from its parent TreeViewItem?

### DISCUSSION THREAD

**Discussion by stackdamage** [AvaloniaUI/Avalonia#15353] [Q&A] [answered]
I am trying to manully add or remove a TreeViewItem from its parent TreeViewItem. In WPF, I can do such operation with following codes. 

           // Get a TreeViewItem
            TreeViewItem parentTI = GetParentTI();
            
            // Append a new child TreeViewItem 
            var aSubTi = new TreeViewItem() { Header = "Sub Child #1" };
            parentTI.Items.Add(aSubTi);

            // Remove a existing child TreeViewItem
            parentTI.Items.Remove(aSubTi);

However, I tried many methods in Avalonia 11, still cannot find a right way to successfuly implement such operation. Sometimes it  throws exception which says 'System.InvalidOperationException: The control already has a visual parent.'.  Is there any path to solve the problem? This question took me a long time. Thanks in advanced if anyone can give any tips.

**Comment by timunie** [comment]
> Thanks in advanced if anyone can give any tips.

Better to use the MVVM pattern and use a TreeDataTemplate, see: https://docs.avaloniaui.net/docs/reference/controls/detailed-reference/treeview-1#example

**Reply by stackdamage** [reply]
Thanks for your reply.  The operation senario is when user right click one of the TreeViewItem, a context menu is shown. Then user click Insert menu, and it's desired to append a new TreeViewItem behid the selected TreeViewItem.  It would be better if the new inserted TreeViewItem can be selected after the insert operation is done. I am not sure if the MVVM pattern can implement such operation. I am still doing research on this.

**Comment by workgroupengineering** [comment, marked as answer]
[TreeViewSample.zip](https://github.com/AvaloniaUI/Avalonia/files/15095029/TreeViewSample.zip)

**Reply by stackdamage** [reply]
I tested your codes. It does work in MVVM pattern. I will learn from the sample and make changes in my project. Thank you @workgroupengineering
